GVFS/GVFS.Platform.Mac/MacFileBasedLock.cs

```diff
@@ -0,0 +1,119 @@
+﻿using GVFS.Common;
+using GVFS.Common.FileSystem;
+using GVFS.Common.Tracing;
+using System;
+using System.Runtime.InteropServices;
+using System.Text;
+
+namespace GVFS.Platform.Mac
+{
+    public class MacFileBasedLock : FileBasedLock
+    {
+        private const int InvalidFileDescriptor = -1;
+
+        private const int LockSh = 1; // #define LOCK_SH   1    /* shared lock */
+        private const int LockEx = 2; // #define LOCK_EX   2    /* exclusive lock */
+        private const int LockNb = 4; // #define LOCK_NB   4    /* don't block when locking */
+        private const int LockUn = 8; // #define LOCK_UN   8    /* unlock */
+
+        private int lockFileDescriptor;
+        private bool lockAcquired;
+
+        public MacFileBasedLock(
+            PhysicalFileSystem fileSystem,
+            ITracer tracer,
+            string lockPath,
+            string signature)
+            : base(fileSystem, tracer, lockPath, signature) 
+        {
+            this.lockFileDescriptor = InvalidFileDescriptor;
+        }
+
+        public override bool TryAcquireLock()
+        {
+            if (this.lockAcquired)
+            {
+                return true;
+            }
+
+            if (this.lockFileDescriptor == InvalidFileDescriptor)
+            {
+                this.lockFileDescriptor = Creat(this.LockPath, Convert.ToInt32("644", 8));
+                if (this.lockFileDescriptor == InvalidFileDescriptor)
+                {
+                    int errno = Marshal.GetLastWin32Error();
+
+                    this.Tracer.RelatedWarning($"Failed to create lock file descriptor for '{this.LockPath}': {errno}"); 
+
+                    return false;
+                }
+            }
+
+            if (Flock(this.lockFileDescriptor, LockEx | LockNb) != 0)
+            {
+                int errno = Marshal.GetLastWin32Error();
+
+                // Log error if not EWOULDBLOCK
+                this.Tracer.RelatedInfo($"Failed to acquire lock for '{this.LockPath}': {errno}");
+
+                return false;
+            }
+
+            byte[] signatureBytes = Encoding.UTF8.GetBytes(this.Signature);
+            long bytesWritten = Write(
+                this.lockFileDescriptor,
+                signatureBytes,
+                Convert.ToUInt64(signatureBytes.Length));
+            
+            if (bytesWritten == -1)
+            {
+                int errno = Marshal.GetLastWin32Error();
+
+                this.Tracer.RelatedWarning($"Failed to write signature for '{this.LockPath}': {errno}");
+            }
+
+            this.Tracer.RelatedInfo($"Lock acquired for for '{this.LockPath}'");
+
+            return true;

```

Comment:
Should we be setting `this.lockAcquired = true` here?

Reply:
Yes we should, good catch :)